### PR TITLE
Event Manager handle_in_client hooks

### PR DESCRIPTION
**Pull request by tokatoka**

### CODE REVIEW

File: libafl/src/events/hooks/mod.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        _fuzzer: &mut Z,
+        _executor: &mut E,
+        _state: &mut S,
+        _client_id: ClientId,
+    ) -> Result<bool, Error> {
+        Ok(true)
+    }
+}
+
+impl<Head, Tail, S> EventManagerHooksTuple<S> for (Head, Tail)
```
Comment (domenukk) [inline libafl/src/events/hooks/mod.rs:85]: The old question: why don't we just do a `vec` with the hooks in it?

Comment (addisoncrump) [reply libafl/src/events/hooks/mod.rs:85]: Cannot constrain per-type/cannot look for specific trait impls/etc.

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: just give it all we got

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: Adding and removing hooks on the fly will be useful

Comment (addisoncrump) [reply libafl/src/events/hooks/mod.rs:85]: I just don't understand why one wouldn't implement a hook with dynamic behaviour instead. You aren't going to be generating hooks.

Comment (tokatoka) [reply libafl/src/events/hooks/mod.rs:85]: can't do this because pre_exec/post_exec have generics <E, Z>

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: FnMut<E, Z> can have generics too

Comment (tokatoka) [reply libafl/src/events/hooks/mod.rs:85]: i mean i can't make them into vec because the compiler says no because it has generics

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: Shoudn't `Vec<Box<dyn FnMut<E, Z>(&blubb) -> bla>>` work?

Comment (addisoncrump) [reply libafl/src/events/hooks/mod.rs:85]: You can't make a generic dyn. You can have a concretised generic type in a dyn, but you cannot have dyns with generic arguments.

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: They are concrete here, right? If they are in an `struct blubb<E, Z>`

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bca6fdf9ae5fa1aa85e0193a6cd47c8d

This is what I meant

Comment (addisoncrump) [reply libafl/src/events/hooks/mod.rs:85]: That's not the situation... The function being called must be generic. There's no concrete type A or B beforehand.

File: libafl/src/events/llmp.rs [hunk 2]
```diff
@@ -593,9 +629,6 @@ where
             } => {
                 log::info!("Received new Testcase from {client_id:?} ({client_config:?}, forward {forward_id:?})");
 
-                if let Ok(meta) = state.metadata_mut::<TransferringMetadata>() {
```
Comment (tokatoka) [inline libafl/src/events/llmp.rs:596]: so you ok with this?
@addisoncrump 

there're two way to achieve what you want now.
1st way is to just make a hook that set this metadata (which i don't recommend)
2nd way is to make a hook that does what you want when new testcase comes, and return `false` then the subsequent testcase processing is skipped

File: libafl/src/events/hooks/mod.rs (hunk 1 above)
Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: Ah I get what you mean... The obvious solution would be not to give the hooks access to executor or fuzzer

Comment (domenukk) [reply libafl/src/events/hooks/mod.rs:85]: In any case, access to state and event manager should be fine tbh

**Pull request merged**